File: elixir/list-ops/lib/list_ops.ex
defmodule ListOps do
  # Please don't use any external modules (especially List or Enum) in your
  # implementation. The point of this exercise is to create these basic
  # functions yourself. You may use basic Kernel functions (like `Kernel.+/2`
  # for adding numbers), but please do not use Kernel functions for Lists like
  # `++`, `--`, `hd`, `tl`, `in`, and `length`.

  @spec count(list) :: non_neg_integer
  def count(l) do
    count(l, 0)
  end

  @spec count(list) :: non_neg_integer
  defp count([], acc), do: acc
  defp count([_h | t], acc), do: count(t, acc + 1)

  @spec reverse(list) :: list
  def reverse(list) do
    reverse(list, [])
  end

  @spec reverse(list, list) :: list
  defp reverse([], rev), do: rev
  defp reverse([h | t], rev), do: reverse(t, [h | rev])

  @spec map(list, (any -> any)) :: list
  def map([], _f), do: []
  def map([h | t], f) do
    [f.(h) | map(t, f)]
  end

  @spec filter(list, (any -> as_boolean(term))) :: list
  def filter([], _f) do
    []
  end

  @spec filter(list, (any -> as_boolean(term))) :: list
  def filter([h | t], f) do
    if f.(h) do
      [h | filter(t, f)]
    else
      filter(t, f)
    end
  end

  @type acc :: any
  @spec reduce(list, acc, (any, acc -> acc)) :: acc
  def reduce([], acc, _f) do
    acc
  end

  @spec reduce(list, acc, (any, acc -> acc)) :: acc
  def reduce([h | t], acc, f) do
    reduce(t, f.(h, acc), f)
  end

  @spec append(list, list) :: list
  def append([], []) do
    []
  end

  @spec append(list, list) :: list
  def append([h|t], l2) do
    [h | append(t, l2)]
  end
  def append([], l2), do: l2

  @spec concat([[any]]) :: [any]
  def concat([]) do
    []
  end

  @spec concat([[any]]) :: [any]
  def concat([h | t]) do
    append(h, concat(t))
  end
end

File: haskell/list-ops/src/ListOps.hs
{-# LANGUAGE BangPatterns #-}

module ListOps
  ( length
  , reverse
  , map
  , filter
  , foldr
  , foldl'
  , (++)
  , concat
  ) where

import Prelude hiding
  ( length, reverse, map, filter, foldr, (++), concat )

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' _ !z [] = z
foldl' f !z (x:xs) = foldl' f (f z x) xs

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr _ z [] = z
foldr f z (x:xs) = f x (foldr f z xs)


length :: [a] -> Int
length = foldr (\_ -> (+) 1) 0

reverse :: [a] -> [a]
reverse xs = reverse' xs []

reverse' :: [a] -> [a] -> [a]
reverse' xs acc = foldl' (flip (:)) acc xs

map :: (a -> b) -> [a] -> [b]
map _ [] = []
map f (x:xs) = f x : map f xs

filter :: (a -> Bool) -> [a] -> [a]
filter _ [] = []
filter p (x:xs) = if p x then x : filter p xs else filter p xs

(++) :: [a] -> [a] -> [a]
[] ++ [] = []
xs ++ [] = xs
[] ++ ys = ys
(x:xs) ++ ys = x:xs ++ ys

concat :: [[a]] -> [a]
concat [] = []
concat [xs] = xs
concat (xs:xss) = xs ++ concat xss
